### PR TITLE
Clear old def-use record when re-analyze an ID

### PR DESCRIPTION
**Pull request by Qining**
This is to fix the bug in #336

### CODE REVIEW

File: source/opt/def_use_manager.cpp [hunk 1]
```diff
@@ -116,6 +105,29 @@ bool DefUseManager::ReplaceAllUsesWith(uint32_t before, uint32_t after) {
   return true;
 }
 
+bool DefUseManager::ClearRecordFor(uint32_t def_id) {
+  if (id_to_def_.count(def_id) == 0) return false;
+
+  // Go through all ids usd by this instruction, remove this instruction's
```
Comment (dneto0) [inline source/opt/def_use_manager.cpp]: "usd" -> "used"

Comment (Qining) [inline source/opt/def_use_manager.cpp]: Fixed in 60b2ab0

File: source/opt/def_use_manager.h [hunk 2]
```diff
@@ -91,6 +91,14 @@ class DefUseManager {
   bool ReplaceAllUsesWith(uint32_t before, uint32_t after);
 
  private:
+  // Clear the internal def-use record of a defined id. Returns true on
+  // success, false if the given |def_id| is not recorded by this manager. This
+  // method erases both the uses of |def_id| and the |def_id|-generating
+  // instruction's use information kept in this manager, but not the operands
+  // in the original instructions. This method does not turn the
```
Comment (dneto0) [inline source/opt/def_use_manager.h]: The last sentence is a little weird. I never would have expected it to mutate the instruction.

File: source/opt/def_use_manager.cpp [hunk 3]
```diff
@@ -43,6 +43,13 @@ void DefUseManager::AnalyzeDefUse(ir::Module* module) {
 
 void DefUseManager::AnalyzeInstDefUse(ir::Instruction* inst) {
   const uint32_t def_id = inst->result_id();
+
+  // If the inst result id has already been recorded in this manager, we need
+  // to clear its old records first then re-analyze its use-def record.
+  if (id_to_def_.count(def_id) != 0) {
```
Comment (dneto0) [inline source/opt/def_use_manager.cpp]: Just call ClearRecordFor since it will do the count check anyway.

File: source/opt/def_use_manager.h [hunk 4]
```diff
@@ -91,6 +91,14 @@ class DefUseManager {
...
+  // |def_id|-generating instruction to nop.
+  bool ClearRecordFor(uint32_t def_id);
```
Comment (dneto0) [inline source/opt/def_use_manager.h]: Better name would be ClearRecordForDefinition or ClearDefinition(uint32_t def_id)

**Comment by dneto0** [comment]
-1

File: source/opt/def_use_manager.h (hunk 4 above)
Comment (Qining) [inline source/opt/def_use_manager.h]: Changed to `ClearDef()` to be consistent with others: `GetDef()`, `KillDef()`.

File: source/opt/def_use_manager.h (hunk 2 above)
Comment (Qining) [inline source/opt/def_use_manager.h]: Last sentence removed in 3087458.

File: source/opt/def_use_manager.cpp (hunk 3 above)
Comment (Qining) [inline source/opt/def_use_manager.cpp]: Fixed in 3087458

**Comment by dneto0** [comment]
+2 please squash first.
Why does ClearDef return a bool when you never use the result?

**Comment by dneto0** [comment]
+2  please squash

**Comment by dneto0** [comment]
oh, there's only one commit.  so it's already squashed. :-)

**Pull request merged**